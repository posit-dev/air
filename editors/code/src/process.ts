import * as cp from "child_process";

export interface CommandResult {
	type: "result";
	code: number | null;
	stdout: string;
	stderr: string;
}

export interface CommandError {
	type: "error";
	error: Error;
}

export function isResult(
	result: CommandResult | CommandError,
): result is CommandResult {
	return result.type == "result";
}

export function isError(
	result: CommandResult | CommandError,
): result is CommandError {
	return result.type == "error";
}

/**
 * Spawns a process and runs a command
 *
 * Collects stdout and stderr emitted along the way.
 *
 * @returns Returns a promise that results when the process exits or errors
 */
export async function runCommand(
	command: string,
	args?: readonly string[],
	options?: cp.SpawnOptionsWithoutStdio,
): Promise<CommandResult | CommandError> {
	return new Promise<CommandResult | CommandError>((resolve) => {
		let stdout = "";
		let stderr = "";

		// Use spawn instead of exec to avoid maxBufferExceeded error
		const p = cp.spawn(command, args, options);

		p.stdout.setEncoding("utf8");
		p.stdout.on("data", (data) => (stdout += data));

		p.stderr.setEncoding("utf8");
		p.stderr.on("data", (data) => (stderr += data));

		// `error` should fire before `close` in the event both fire.
		// `error` should also fire no matter what.
		p.on("error", (error) => {
			return resolve({
				type: "error",
				error: error,
			});
		});

		// Use `close` to wait for stdio streams to close too
		p.on("close", (code) => {
			return resolve({
				type: "result",
				code: code,
				stdout: stdout,
				stderr: stderr,
			});
		});
	});
}
