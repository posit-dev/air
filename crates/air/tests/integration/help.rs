use std::process::Command;

use crate::helpers::binary_path;
use crate::helpers::CommandExt;

#[test]
fn test_help() {
    insta::assert_snapshot!(&mut Command::new(binary_path())
        .run()
        .normalize_os_executable_name());
    insta::assert_snapshot!(Command::new(binary_path())
        .arg("help")
        .run()
        .normalize_os_executable_name());
    insta::assert_snapshot!(Command::new(binary_path())
        .arg("--help")
        .run()
        .normalize_os_executable_name());
    insta::assert_snapshot!(Command::new(binary_path())
        .arg("-h")
        .run()
        .normalize_os_executable_name());
}

#[test]
fn test_format_help() {
    insta::assert_snapshot!(Command::new(binary_path())
        .arg("format")
        .run()
        .normalize_os_executable_name());
    insta::assert_snapshot!(Command::new(binary_path())
        .arg("format")
        .arg("--help")
        .run()
        .normalize_os_executable_name());
    insta::assert_snapshot!(Command::new(binary_path())
        .arg("format")
        .arg("-h")
        .run()
        .normalize_os_executable_name());
}
