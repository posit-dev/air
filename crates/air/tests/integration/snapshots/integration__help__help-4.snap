---
source: crates/air/tests/integration/help.rs
expression: "Command::new(binary_path()).arg(\"-h\").run().normalize_os_executable_name()"
---
success: true
exit_code: 0
----- stdout -----
Air: An R language server and formatter

Usage: air [OPTIONS] <COMMAND>

Commands:
  format           Format a set of files or directories
  language-server  Start a language server
  help             Print this message or the help of the given subcommand(s)

Options:
  -h, --help     Print help
  -V, --version  Print version

Global options:
      --log-level <LOG_LEVEL>  The log level. One of: `error`, `warn`, `info`, `debug`, or `trace`.
                               Defaults to `warn`
      --no-color               Disable colored output. To turn colored output off, either set this
                               option or set the environment variable `NO_COLOR` to any non-zero
                               value

For help with a specific command, see: `air help <command>`.

----- stderr -----

----- args -----
-h
